File: examples/tests/javascript/example.test.js
/**
 * Example test file demonstrating Jest best practices.
 *
 * Tests follow the AAA pattern:
 * - Arrange: Set up test data and preconditions
 * - Act: Execute the code being tested
 * - Assert: Verify the expected outcome
 */

// Simple unit test
describe('Basic Math Operations', () => {
  test('addition works correctly', () => {
    // Arrange
    const a = 2;
    const b = 3;
    const expected = 5;

    // Act
    const result = a + b;

    // Assert
    expect(result).toBe(expected);
  });

  test('multiplication works correctly', () => {
    expect(5 * 3).toBe(15);
  });
});

// Test suite grouping
describe('String Operations', () => {
  test('converts to uppercase', () => {
    const text = 'claude code';
    expect(text.toUpperCase()).toBe('CLAUDE CODE');
  });

  test('converts to lowercase', () => {
    const text = 'CLAUDE CODE';
    expect(text.toLowerCase()).toBe('claude code');
  });

  test('checks string contains substring', () => {
    const text = 'Claude Code Starter';
    expect(text).toContain('Code');
    expect(text).toMatch(/Starter$/);
  });
});

// Parametrized tests (test.each)
describe('Parametrized Tests', () => {
  test.each([
    [1, 2],
    [2, 4],
    [3, 6],
    [10, 20],
    [-5, -10],
  ])('doubles %i to get %i', (input, expected) => {
    expect(input * 2).toBe(expected);
  });
});

// Testing with setup and teardown
describe('Setup and Teardown', () => {
  let data;

  beforeEach(() => {
    // Runs before each test
    data = [];
  });

  afterEach(() => {
    // Runs after each test
    data = null;
  });

  beforeAll(() => {
    // Runs once before all tests in this suite
    console.log('Starting test suite');
  });

  afterAll(() => {
    // Runs once after all tests in this suite
    console.log('Finished test suite');
  });

  test('can push to array', () => {
    data.push(1);
    expect(data).toHaveLength(1);
    expect(data[0]).toBe(1);
  });

  test('starts with empty array', () => {
    expect(data).toHaveLength(0);
    expect(data).toEqual([]);
  });
});

// Testing objects and arrays
describe('Object and Array Matching', () => {
  test('object equality', () => {
    const user = {
      name: 'Test User',
      email: 'test@example.com',
      age: 25
    };

    expect(user).toEqual({
      name: 'Test User',
      email: 'test@example.com',
      age: 25
    });
  });

  test('object contains properties', () => {
    const user = { name: 'Alice', age: 30, role: 'admin' };

    expect(user).toHaveProperty('name');
    expect(user).toHaveProperty('age', 30);
    expect(user).toMatchObject({ name: 'Alice' });
  });

  test('array operations', () => {
    const numbers = [1, 2, 3, 4, 5];

    expect(numbers).toHaveLength(5);
    expect(numbers).toContain(3);
    expect(numbers).toEqual(expect.arrayContaining([2, 4]));
  });
});

// Testing exceptions
describe('Exception Handling', () => {
  test('throws error on division by zero', () => {
    const divideByZero = () => {
      throw new Error('Division by zero');
    };

    expect(divideByZero).toThrow();
    expect(divideByZero).toThrow(Error);
    expect(divideByZero).toThrow('Division by zero');
  });

  test('does not throw with valid input', () => {
    const safeDivide = () => 10 / 2;
    expect(safeDivide).not.toThrow();
  });
});

// Testing async code
describe('Async Operations', () => {
  test('promise resolves correctly', async () => {
    const fetchData = () => Promise.resolve('data');
    const data = await fetchData();
    expect(data).toBe('data');
  });

  test('promise rejects correctly', async () => {
    const fetchError = () => Promise.reject(new Error('Failed'));
    await expect(fetchError()).rejects.toThrow('Failed');
  });

  test('async function with timeout', async () => {
    const delay = (ms) => new Promise(resolve => setTimeout(resolve, ms));
    const start = Date.now();
    await delay(100);
    const elapsed = Date.now() - start;
    expect(elapsed).toBeGreaterThanOrEqual(100);
  });
});

// Mocking functions
describe('Mocking', () => {
  test('mock function calls', () => {
    const mockFn = jest.fn();
    mockFn('arg1', 'arg2');

    expect(mockFn).toHaveBeenCalled();
    expect(mockFn).toHaveBeenCalledTimes(1);
    expect(mockFn).toHaveBeenCalledWith('arg1', 'arg2');
  });

  test('mock function return values', () => {
    const mockFn = jest.fn()
      .mockReturnValueOnce(10)
      .mockReturnValueOnce(20)
      .mockReturnValue(30);

    expect(mockFn()).toBe(10);
    expect(mockFn()).toBe(20);
    expect(mockFn()).toBe(30);
    expect(mockFn()).toBe(30);
  });

  test('mock implementation', () => {
    const mockFn = jest.fn((x, y) => x + y);

    expect(mockFn(2, 3)).toBe(5);
    expect(mockFn).toHaveBeenCalledWith(2, 3);
  });
});

// Snapshot testing
describe('Snapshot Testing', () => {
  test('matches snapshot', () => {
    const data = {
      name: 'Claude Code',
      version: '1.0.0',
      features: ['hooks', 'commands', 'skills']
    };

    expect(data).toMatchSnapshot();
  });
});

// Testing with specific matchers
describe('Advanced Matchers', () => {
  test('number comparisons', () => {
    const value = 100;

    expect(value).toBeGreaterThan(50);
    expect(value).toBeLessThan(200);
    expect(value).toBeGreaterThanOrEqual(100);
    expect(value).toBeLessThanOrEqual(100);
  });

  test('floating point equality', () => {
    const value = 0.1 + 0.2;
    // Don't use toBe for floats
    expect(value).toBeCloseTo(0.3);
  });

  test('truthy and falsy', () => {
    expect(true).toBeTruthy();
    expect(1).toBeTruthy();
    expect('text').toBeTruthy();

    expect(false).toBeFalsy();
    expect(0).toBeFalsy();
    expect('').toBeFalsy();
    expect(null).toBeFalsy();
    expect(undefined).toBeFalsy();
  });

  test('null and undefined', () => {
    const n = null;
    expect(n).toBeNull();
    expect(n).toBeDefined();
    expect(n).not.toBeUndefined();

    let u;
    expect(u).toBeUndefined();
    expect(u).not.toBeNull();
    expect(u).not.toBeDefined();
  });
});

// Test skipping and focusing
describe('Test Control', () => {
  test('this test runs', () => {
    expect(true).toBe(true);
  });

  test.skip('this test is skipped', () => {
    expect(false).toBe(true); // Won't fail
  });

  // Use test.only to run only this test (useful for debugging)
  // test.only('only this test runs', () => {
  //   expect(true).toBe(true);
  // });
});

// User validation example
describe('UserService', () => {
  class UserService {
    validateEmail(email) {
      return email && email.includes('@') && email.includes('.');
    }

    normalizeEmail(email) {
      return email.trim().toLowerCase();
    }

    isAdult(age) {
      return age >= 18;
    }
  }

  let userService;

  beforeEach(() => {
    userService = new UserService();
  });

  describe('validateEmail', () => {
    test('accepts valid email', () => {
      expect(userService.validateEmail('user@example.com')).toBe(true);
    });

    test('rejects email without @', () => {
      expect(userService.validateEmail('userexample.com')).toBe(false);
    });

    test('rejects email without domain', () => {
      expect(userService.validateEmail('user@')).toBe(false);
    });

    test('rejects empty email', () => {
      expect(userService.validateEmail('')).toBe(false);
    });
  });

  describe('normalizeEmail', () => {
    test('trims whitespace', () => {
      expect(userService.normalizeEmail('  user@example.com  '))
        .toBe('user@example.com');
    });

    test('converts to lowercase', () => {
      expect(userService.normalizeEmail('USER@EXAMPLE.COM'))
        .toBe('user@example.com');
    });
  });

  describe('isAdult', () => {
    test.each([
      [18, true],
      [25, true],
      [100, true],
      [17, false],
      [0, false],
      [-1, false],
    ])('age %i returns %s', (age, expected) => {
      expect(userService.isAdult(age)).toBe(expected);
    });
  });
});

File: examples/tests/go/example_test.go
// Example test file for Go
// Test files must end with _test.go
// Test functions must start with Test and take *testing.T

package example

import (
	"errors"
	"strings"
	"testing"
)

// Simple functions to test
func Add(a, b int) int {
	return a + b
}

func Multiply(a, b int) int {
	return a * b
}

func Divide(a, b int) (int, error) {
	if b == 0 {
		return 0, errors.New("division by zero")
	}
	return a / b, nil
}

// User struct for testing
type User struct {
	Name  string
	Email string
	Age   int
}

func NewUser(name, email string, age int) *User {
	return &User{
		Name:  name,
		Email: email,
		Age:   age,
	}
}

func (u *User) IsAdult() bool {
	return u.Age >= 18
}

func (u *User) ValidateEmail() bool {
	return strings.Contains(u.Email, "@") && strings.Contains(u.Email, ".")
}

// ===== TESTS =====

// Basic test
func TestAdd(t *testing.T) {
	result := Add(2, 3)
	expected := 5

	if result != expected {
		t.Errorf("Add(2, 3) = %d; want %d", result, expected)
	}
}

func TestMultiply(t *testing.T) {
	result := Multiply(2, 3)
	expected := 6

	if result != expected {
		t.Errorf("Multiply(2, 3) = %d; want %d", result, expected)
	}
}

// Test with subtests
func TestDivide(t *testing.T) {
	t.Run("success", func(t *testing.T) {
		result, err := Divide(10, 2)
		if err != nil {
			t.Errorf("Unexpected error: %v", err)
		}
		if result != 5 {
			t.Errorf("Divide(10, 2) = %d; want 5", result)
		}
	})

	t.Run("division by zero", func(t *testing.T) {
		_, err := Divide(10, 0)
		if err == nil {
			t.Error("Expected error for division by zero, got nil")
		}
	})
}

// Table-driven tests (Go idiom)
func TestAddTableDriven(t *testing.T) {
	tests := []struct {
		name     string
		a        int
		b        int
		expected int
	}{
		{"positive numbers", 2, 3, 5},
		{"negative numbers", -2, -3, -5},
		{"mixed signs", -2, 3, 1},
		{"zero", 0, 0, 0},
		{"large numbers", 1000, 2000, 3000},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := Add(tt.a, tt.b)
			if result != tt.expected {
				t.Errorf("Add(%d, %d) = %d; want %d",
					tt.a, tt.b, result, tt.expected)
			}
		})
	}
}

// Test User struct
func TestUserCreation(t *testing.T) {
	user := NewUser("Alice", "alice@example.com", 25)

	if user.Name != "Alice" {
		t.Errorf("user.Name = %s; want Alice", user.Name)
	}
	if user.Email != "alice@example.com" {
		t.Errorf("user.Email = %s; want alice@example.com", user.Email)
	}
	if user.Age != 25 {
		t.Errorf("user.Age = %d; want 25", user.Age)
	}
}

func TestUserIsAdult(t *testing.T) {
	tests := []struct {
		name     string
		age      int
		expected bool
	}{
		{"adult", 18, true},
		{"over 18", 25, true},
		{"minor", 17, false},
		{"child", 5, false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			user := NewUser("Test", "test@example.com", tt.age)
			result := user.IsAdult()
			if result != tt.expected {
				t.Errorf("IsAdult() with age %d = %v; want %v",
					tt.age, result, tt.expected)
			}
		})
	}
}

func TestUserValidateEmail(t *testing.T) {
	tests := []struct {
		name     string
		email    string
		expected bool
	}{
		{"valid email", "user@example.com", true},
		{"no at symbol", "userexample.com", false},
		{"no domain", "user@", false},
		{"no local part", "@example.com", false},
		{"empty", "", false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			user := NewUser("Test", tt.email, 25)
			result := user.ValidateEmail()
			if result != tt.expected {
				t.Errorf("ValidateEmail(%s) = %v; want %v",
					tt.email, result, tt.expected)
			}
		})
	}
}

// Helper function example
func assertEqual(t *testing.T, got, want interface{}) {
	t.Helper() // Marks this as a helper function
	if got != want {
		t.Errorf("got %v; want %v", got, want)
	}
}

func TestWithHelper(t *testing.T) {
	result := Add(2, 3)
	assertEqual(t, result, 5)
}

// Benchmark example
func BenchmarkAdd(b *testing.B) {
	for i := 0; i < b.N; i++ {
		Add(2, 3)
	}
}

func BenchmarkMultiply(b *testing.B) {
	for i := 0; i < b.N; i++ {
		Multiply(2, 3)
	}
}

// Example test (appears in documentation)
func ExampleAdd() {
	result := Add(2, 3)
	println(result)
	// Output: 5
}

// Parallel test example
func TestParallel(t *testing.T) {
	t.Run("group", func(t *testing.T) {
		t.Run("test1", func(t *testing.T) {
			t.Parallel()
			// Test runs in parallel with test2
			result := Add(1, 1)
			assertEqual(t, result, 2)
		})

		t.Run("test2", func(t *testing.T) {
			t.Parallel()
			// Test runs in parallel with test1
			result := Add(2, 2)
			assertEqual(t, result, 4)
		})
	})
}

// Test with cleanup
func TestWithCleanup(t *testing.T) {
	// Setup
	data := make([]int, 0)

	// Cleanup will run after test completes
	t.Cleanup(func() {
		data = nil
	})

	// Test
	data = append(data, 1, 2, 3)
	if len(data) != 3 {
		t.Errorf("len(data) = %d; want 3", len(data))
	}
}

// Skip test conditionally
func TestSkipped(t *testing.T) {
	if testing.Short() {
		t.Skip("Skipping test in short mode")
	}
	// Long-running test code here
}
